locale/en/docs/guides/dont-block-the-event-loop.md

```diff
@@ -0,0 +1,435 @@
+```javascript
+function asyncAvg(n, avgCB) {
+  // Save ongoing sum in JS closure.
+  var sum = 0;
+  function help(i, cb) {
+    sum += i;
+    if (i == n) {
+      cb(sum);
+      return;
+    }
+
+    // "Asynchronous recursion".
+    // Schedule next operation asynchronously.
+    setImmediate(help.bind(null, i+1, cb));
+  }
+
+  // Start the helper, with CB to call avgCB.
+  help(1, function(sum){
+      var avg = sum/n;
+      avgCB(avg);
+  });
+}
+
+avg(n, function(avg){
+  console.log('avg of 1-n: ' + avg);
+});
+```
+
+You can apply this principle to array iterations and so forth.
+
+#### Offloading
+If you need to do something more complex, partitioning is not a good option.
+This is because partitioning uses only the Event Loop, and you won't benefit from multiple cores almost certainly available on your machine.
+*Remember, the Event Loop should orchestrate client requests, not fulfill them itself.*
+For a complicated task, move the work off of the Event Loop onto a Worker Pool.
+
+##### How to offload
+You have two options for a destination Worker Pool to which to offload work.
+1. You can use the built-in Node Worker Pool by developing a [C++ addon](https://nodejs.org/api/addons.html). On older versions of Node, build your C++ addon using [NAN](https://github.com/nodejs/nan), and on newer versions use [N-API](https://nodejs.org/api/n-api.html). [node-webworker-threads](https://www.npmjs.com/package/webworker-threads) offers a JavaScript-only way to access Node's Worker Pool.
+2. You can create and manage your own Worker Pool dedicated to computation rather than Node's I/O-themed Worker Pool. The most straightforward ways to do this is using [Child Process](https://nodejs.org/api/child_process.html) or [Cluster](https://nodejs.org/api/cluster.html).
+
+You should *not* simply create a [Child Process](https://nodejs.org/api/child_process.html) for every client.
+You can receive client requests more quickly than you can create and manage children, and your server would then be essentially a [fork bomb](https://en.wikipedia.org/wiki/Fork_bomb).
+
+##### Downside of offloading
+The downside of the offloading approach is that it incurs overhead in the form of *communication costs*.
+Only the Event Loop is allowed to see the "namespace" (JavaScript state) of your application.
+From a Worker, you cannot manipulate a JavaScript object in the Event Loop's namespace.
+Instead, you have to serialize and deserialize any objects you wish to share.
+Then the Worker can operate on its own copy of these object(s) and return the modified object (or a "patch") to the Event Loop.
+
+For serialization concerns, see the section on JSON DOS.
+
+##### Best practices for offloading
+Follow these best practices for offloading:
+1. You should create your own Computation Worker Pool for extensive computation.
+2. The Computation Worker Pool should have no more than Workers than your machine has [logical cores](https://nodejs.org/api/os.html#os_os_cpus).
```

Comment:
what about when you have multiple addons, each with a number of threads equal to the number of logical cores?

Reply:
I would suggest creating a single Computation Worker Pool at the application level, similar to the Node Worker Pool that handles I/O from all points in the application. Do you have any reference applications I can look at to describe alternate patterns?

Reply:
but no such thing exists, so current addons don't use it. it has been proposed that uv expose two work APIs, one for blocking non-cpu intensive calls (file i/o, dns.lookup), the other for cpu intensive activities (crypto.pbkdf for key generation, etc), but no code has been PRed, or benchmarks to show it will work. so, in the meantime, you have to consider all the internal pools of all the addons, and actually most addons will mix cpu and i/o intensive work in the same uv work pool. I'm slowly working towards getting metrics in node about the state of the uv pool, but that doesn't exist yet, so not worth mentioning here.

Reply:
OK. If I had added such an API would it be of interest? Is anyone working on this right now?

Reply:
Also, I'll just remove this section. It sounds like this is too much of an open problem to worry about right now.